### DISCUSSION TITLE
How to combine CommandBuffer with PassNode?

### DISCUSSION THREAD

**Discussion by FuXiii** [skaarj1989/FrameGraph#5] [Q&A] [answered]
Hello I am coming again~

Before run `PassNode::Execute()`, you should create the image resource that it need, and after it you destroy them which it didn't need.

Combine with `CommnadBuffer` there has two case:
1. Each frame have one `CommandBuffer`, in another word each `FrameGraph` have one `CommandBuffer`
2. Each `PassNode` have one `CommandBuffer`

Before you `Submit` your `CommandBuffer` you shouldn't destroy the image resource.
So if use `CommandBuffer` with case `1`, you shouldn't destroy the image resource after `PassNode::Execute()`, because the `CommandBuffer` will `Submit` until all the `PassNode` had executed.
So if use `CommandBuffer` with case `2`, you should create `CommandBuffer` each `PassNode` and `Submit` the `CommandBuffer` and wait the `CommandBuffer` finished before destroy the resources

Which one may be elegant or reasonable？
How to combine with `CommandBuffer` make them reasonable?

**Comment by skaarj1989** [comment, marked as answer]
Hi @FuXiii 

I'm sorry for the late reply (due to sickness).

Here is my advice:
- You need a transient resource system (see: [#1](https://github.com/skaarj1989/FrameGraph-Example/blob/eb884e8945b20189c088a6893cc5089396c44562/src/TransientResources.cpp) and [#2](https://github.com/skaarj1989/FrameGraph-Example/blob/master/src/FrameGraphTexture.cpp))
- Pass your `CommandBuffer` to a FrameGraph (see: [WorldRenderer](https://github.com/skaarj1989/FrameGraph-Example/blob/eb884e8945b20189c088a6893cc5089396c44562/src/WorldRenderer.cpp#L270)) and use it in a pass execution lambda (like in here: [TonemapPass](https://github.com/skaarj1989/FrameGraph-Example/blob/eb884e8945b20189c088a6893cc5089396c44562/src/Passes/TonemapPass.cpp#L60)).

**Comment by FuXiii** [comment]
Sorry for replying so late too Σ( ° △ °|||)︴
Yes, the really Image resource create in `FrameGraph::Execute` stage. It will use `context` to call `RenderContext::createTexture2D(...)` and final call `glCreateTextures(...)` in `RenderContext::_createImmutableTexture(...)`

**Comment by FuXiii** [comment]
Before to create image or buffer, I need to design `RenderContext`, after that I can use it to create resource.（づ￣3￣）づ
